apollo-spaceport/src/server.rs

```diff
@@ -0,0 +1,328 @@
+        let (tx, mut rx) = tokio::sync::mpsc::channel::<()>(10);
+        let total = Arc::new(AtomicU32::new(0u32));
+        let task_total = total.clone();
+        tokio::task::spawn(async move {
+            let client = Client::new();
+            let mut interval = interval(INGRESS_CLOCK_TICK);
+            interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+            interval.tick().await;
+            loop {
+                tokio::select! {
+                    biased;
+                    mopt = rx.recv() => {
+                        tracing::debug!("spaceport triggered");
+                        match mopt {
+                            Some(_msg) => process_all_graphs(&client, task_graph_usage.clone()).await,
+                            None => break
+                        }
+                    },
+                    _ = interval.tick() => {
+                        tracing::debug!("spaceport ticked");
+                        task_total.store(0, Ordering::SeqCst);
+                        process_all_graphs(&client, task_graph_usage.clone()).await;
+                    }
+                };
+            }
+        });
+        Self {
+            addr,
+            graph_usage,
+            tx,
+            total,
+        }
+    }
+
+    /// Start serving requests.
+    pub async fn serve(self) -> Result<(), Error> {
+        let addr = self.addr;
+        Server::builder()
+            .add_service(ReporterServer::new(self))
+            .serve(addr)
+            .await
+    }
+
+    async fn submit_report(
+        client: &Client,
+        key: String,
+        report: Report,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        tracing::debug!("submitting report: {:?}", report);
+        // Protobuf encode message
+        let mut content = BytesMut::new();
+        report
+            .encode(&mut content)
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+        // Create a gzip encoder
+        let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+        // Write our content to our encoder
+        encoder
+            .write_all(&content)
+            .map_err(|e| Status::internal(e.to_string()))?;
+        // Finish encoding and retrieve content
+        let compressed_content = encoder
+            .finish()
+            .map_err(|e| Status::internal(e.to_string()))?;
+        let mut backoff = Duration::from_millis(0);
+        let ingress = match std::env::var("APOLLO_USAGE_REPORTING_INGRESS_URL") {
+            Ok(v) => v,
+            Err(_e) => DEFAULT_APOLLO_USAGE_REPORTING_INGRESS_URL.to_string(),
+        };
+        let req = client
+            .post(ingress)
+            .body(compressed_content)
+            .header("X-Api-Key", key)
+            .header("Content-Encoding", "gzip")
+            .header("Content-Type", "application/protobuf")
+            .header("Accept", "application/json")
```

Comment:
Let's send some sort of user-agent (TS just sends 'ApolloServerPluginUsageReporting')

Reply:
I'm assuming we don't want to masquerade as the same user agent since it might be useful to distinguish who is sending data to the ingress.  How about:
`ApolloRouterPluginUsageReporting`
?

Reply:
eh, it's not really a plugin? i think it would be fine to say `apollo-spaceport` or something.